### DISCUSSION TITLE
Http Client Pool and middleware

### DISCUSSION THREAD

**Discussion by vitsw** [laravel/framework#40760] [Q&A]
I have to push a log middleware to the asynchronous requests. When I add a request to the Pool it initializes PendingRequest by calling setHandler() method and async() method. It means that the client is build, all handlers are pushed to the handlerStack and client is marked as reusable. So no middleware will be pushed to the handlerStack afterwards.

For example, those middleware won't be pushed to the handlerStack:

```php
$responses = Http::pool(fn (Pool $pool) => [
    $pool->as('first')->withMiddleware(Middleware::log())->post('https://test1.com'),
    $pool->as('second')->withMiddleware(Middleware::log())->post('https://test2.com')
]);
```

Is it intended behavior? Is there some pitfalls with using middleware with async requests? Or it works in this way just because it was unnecessary feature?

**Comment by ousid** [comment]
Hi, @vitsw 

Maybe you need to set the middleware before polling the request.?

```php
$responses = Http::withMiddleware(Middleware::log)->poll(fn (Pool $pool) => [
    $pool->as('first')->post('https://test1.com'),
    $pool->as('second')->post('https://test2.com')
]);
```

**Reply by vitsw** [reply]
Hi, @ousid. Thanks for the reply. But this does not work either.

**Reply by ousid** [reply]
Hmm, are you sure that you registered the middleware in `kernel.php` file?

**Reply by vitsw** [reply]
It's a [Guzzle Middleware](https://docs.guzzlephp.org/en/stable/handlers-and-middleware.html) not the one from the framework. It augments the functionality of handlers. For instance, I need to log every external request and response.

**Reply by ousid** [reply]
aaa yeah, I'm not familiar with Guzzle middlewares, thanks for the info Schwetz

**Reply by nevmerzhitsky** [reply]
The `Pool` class doesn't deliver `$middleware` content of the `Http` client instance while working with an array of requests, it only deliver a `factory` value:

```php
    public function pool(callable $callback)
    {
        $results = [];

        $requests = tap(new Pool($this->factory), $callback)->getRequests();

        foreach ($requests as $key => $item) {
            $results[$key] = $item instanceof static ? $item->getPromise()->wait() : $item->wait();
        }

        return $results;
    }
```

Thus, any middleware before the `pool()` call "disappears" for the pool instance.

I treat this as a Laravel bug.

**Comment by vitsw** [comment]
When Http::pool($callback) is called it instantiates an instance of PendingRequest and calls its `pool` method passing the `$callback` through:

https://github.com/laravel/framework/blob/c275cdc018b70b87f81f9412ff4287f3d9f51760/src/Illuminate/Http/Client/Factory.php#L390-L392

The `pool` method instantiates the `Pool instance` and calls the `$callback` passing the `Pool instance` as argument

https://github.com/laravel/framework/blob/c275cdc018b70b87f81f9412ff4287f3d9f51760/src/Illuminate/Http/Client/PendingRequest.php#L633

So here we can add requests to the Pool
```php
$responses = Http::withMiddleware(Middleware::log)->poll(fn (Pool $pool) => [
    $pool->as('first')->post('https://test1.com'),
    $pool->as('second')->post('https://test2.com')
]);
```

Pool instantiates new PendingRequest using `asyncRequest()` method which calls `setHandler()` and `async()` methods on it. After that pushes it to the internal $pool array.  

https://github.com/laravel/framework/blob/c275cdc018b70b87f81f9412ff4287f3d9f51760/src/Illuminate/Http/Client/Pool.php#L56-L69

`setHandler()` method creates a guzzle client providing configured handlerStack

https://github.com/laravel/framework/blob/c275cdc018b70b87f81f9412ff4287f3d9f51760/src/Illuminate/Http/Client/PendingRequest.php#L1088-L1095

This is how handlerStack is configured. Some internal handler are pushed to the stack as well as middleware. Particulary when we add a request to the pool the setHandler() method is called on a fresh instance of the PendingRequest so the middleware collection is empty.

https://github.com/laravel/framework/blob/c275cdc018b70b87f81f9412ff4287f3d9f51760/src/Illuminate/Http/Client/PendingRequest.php#L860-L871

After that `async()` method is called. It marks PendingRequest as async.
https://github.com/laravel/framework/blob/c275cdc018b70b87f81f9412ff4287f3d9f51760/src/Illuminate/Http/Client/PendingRequest.php#L1012-L1017

So at this point PendingRequest is configured (client is build and handlerStack is set up), marked as async and added to the Pool.

---

When Pool runs requests `sendRequest()` method is called on every request in a Pool. It builds guzzle client using `buildClient()` method and calls some method on it (ex. get,post etc.)

https://github.com/laravel/framework/blob/c275cdc018b70b87f81f9412ff4287f3d9f51760/src/Illuminate/Http/Client/PendingRequest.php#L742-L754


So when request was marked async, it would reuse client which had been already created and configured in this case using setHandler() method. That means if we add middleware to the request they won't be pushed to the handlerStack.
https://github.com/laravel/framework/blob/c275cdc018b70b87f81f9412ff4287f3d9f51760/src/Illuminate/Http/Client/PendingRequest.php#L803-L828


---

I've solved this by extending the Pool, but I am curious about the reasons it was designed in this way. Or may be I missed something.

**Reply by aliwesome** [reply]
@vitsw Could you please share you workaround for extending Pool class?

**Comment by madelson** [comment]
Will this be fixed? It seems like middleware should be able to work with pooled requests...

**Reply by nevmerzhitsky** [reply]
A workaround is simple: attach the middlewares inside the pool function callback to the `$pool->as('...')` call.

**Reply by madelson** [reply]
@nevmerzhitsky sorry can you explain/link a code snippet of what you mean

**Comment by aliwesome** [comment]
@vitsw Could you please share you workaround for extending Pool class?

**Comment by chris-lee-lb** [comment]
My workaround for reference. (not elegant mayble)

```php
$responses = Http::pool(function (Pool $pool) {
    $decorator = fn (Pool $pool, string $id): PendingRequest => $pool
        ->as($id)
        ->timeout(10)
        ->retry(3, 100, fn ($ex) => $ex instanceof ConnectionException)
        ->withMiddleware('xxx.xxx');

    return [
        $decorator($pool, 'order-001')->get('https://dummyjson.com/c/xxx1'),
        $decorator($pool, 'order-002')->get('https://dummyjson.com/c/xxx2'),
        $decorator($pool, 'order-003')->get('https://dummyjson.com/c/xxx3'),
        $decorator($pool, 'order-004')->get('https://dummyjson.com/c/xxx4'),
        $decorator($pool, 'order-005')->get('https://dummyjson.com/c/xxx5'),
    ];
});
```
